core/src/main/scala/zio/schema/Schema.scala

```diff
@@ -27,6 +27,410 @@ object Schema {
   sealed case class Tuple[A, B](left: Schema[A], right: Schema[B])       extends Schema[(A, B)]
   final case class EitherSchema[A, B](left: Schema[A], right: Schema[B]) extends Schema[Either[A, B]]
 
+  sealed trait CaseClass[Z] extends Schema[Z] {
+    def toRecord: Record
+  }
+  case class CaseClass1[A, Z](field: (String, Schema[A]), construct: A => Z, extract: Z => Tuple1[A])
+      extends CaseClass[Z] {
+    override def toRecord: Record = Record(Map(field))
+  }
+  case class CaseClass2[A1, A2, Z](
+    field1: (String, Schema[A1]),
+    field2: (String, Schema[A2]),
+    construct: (A1, A2) => Z,
+    extract: Z => (A1, A2)
+  ) extends CaseClass[Z] {
+    override def toRecord: Record = Record(Map(field1, field2))
+  }
+
+  case class CaseClass3[A1, A2, A3, Z](
+    field1: (String, Schema[A1]),
+    field2: (String, Schema[A2]),
+    field3: (String, Schema[A3]),
+    construct: (A1, A2, A3) => Z,
+    extract: Z => (A1, A2, A3)
+  ) extends CaseClass[Z] {
```

Comment:
I think this encoding is almost perfect. After studying it for a while, I can see that `extract` will necessarily allocate a tuple. We can eliminate this overhead by switching to a slightly different encoding:

```scala
 case class CaseClass3[A1, A2, A3, Z](
     field1: (String, Schema[A1]),
     field2: (String, Schema[A2]),
     field3: (String, Schema[A3]),
     construct: (A1, A2, A3) => Z,
     extractField1: Z => A1,
     extractField2: Z => A2,
     extractField3: Z => A3
   ) extends CaseClass[Z] 
```

Then we could pick apart the `Z` into its component pieces without doing any allocations; and we can already, having all the component pieces, construct the `Z` by using `construct` without any allocations.

Reply:
I like this a lot. Should set us up well to implement Optics as well.